Add Batch JSON-RPC support (rpc client & server)

jakesyl opened the pull request:
This pr aims to address #329 which has been open for a while now by adding bulk request support. This is based on from [decred#841](https://github.com/decred/dcrd/pull/841) by @davecgh 

Note: This PR was previously accidentally merged in #1503 and reverted in #1549 (followed by a rebase to clean up the commit history) and as a result, there is considerable discussion at #1503

In regards to the multiple commits: once we have consensus on the implementation I'll rebase into a single commit

jakesyl commented:
Rebased!

jakesyl commented:
Also - adding some changes from #1503. There were like 30 commits before so I wanted to squash, but I'll rebase again when we want to merge

jakesyl commented:
Updated with constants as per requests on the last pr

jakesyl commented:
@everyone sorry about the fractal nature, trying to address everyone's reviews on the old pr. 

@torkelrogstad curious why the custom type here would be better than a string. This allows us to use new versions without modifying the code:

![image](https://user-images.githubusercontent.com/6392429/91641112-570b4280-e9f0-11ea-8098-89198970f0d0.png)

Happy to implement this if there's a reason to. If we did, I'd like a validator that we can call to make sure the string conforms to the type. Something like 

```
(rpcVersion) isValid(){
  // check if one of valid versions
}
```

jakesyl commented:
also @jcvernaleo @onyb @torkelrogstad @Roasbeef I just moved all the reviewers over from #1503 feel free to remove

jakesyl commented:
@torkelrogstad implemented your change in 4a97be317a5b4794333f347297349267c0272eb1, I can revert if we want to. I still think if this is done `method` should be a custom type as well

jakesyl commented:
I'll go through and resolve the conflicts this weekend

onyb commented on rpcclient/infrastructure.go:
```diff
@@ -1524,3 +1557,65 @@ func (c *Client) BackendVersion() (BackendVersion, error) {
+		marshalledRequest = append(marshalledRequest, request.marshalledJSON...)
+		marshalledRequest = append(marshalledRequest, []byte(",")...)
+	}
+	if len(marshalledRequest) > 0 {
+		marshalledRequest = marshalledRequest[:len(marshalledRequest)-1]
+	}
```
Deserves a comment explaining that this piece of code removes the last hanging `,` character.

jakesyl replied:
fixed in 9c315730

onyb commented on rpcclient/infrastructure.go:
```diff
@@ -1524,3 +1557,65 @@ func (c *Client) BackendVersion() (BackendVersion, error) {
+	return responseChan
+}
+
+// Send sends batch requests
```
Needs some explanation on what this does, especially stating that this function sends out the actual RPC request over the wire.

jakesyl replied:
Attempted to clarify in: 9c315730

onyb commented on rpcserver.go:
```diff
@@ -3900,7 +3909,60 @@ func createMarshalledReply(id, result interface{}, replyErr error) ([]byte, erro
+		if parsedCmd.err != nil {
+			jsonErr = parsedCmd.err
+		} else {
+			result, _ = s.standardCmdResult(parsedCmd,
+				closeChan)
```
Unhandled error?

jakesyl replied:
good catch

jakesyl replied:
fixed in 9c315730

onyb commented on rpcserver.go:
```diff
@@ -3945,80 +4007,165 @@ func (s *rpcServer) jsonRPCRead(w http.ResponseWriter, r *http.Request, isAdmin
+				Message: fmt.Sprintf("Failed to parse request: %v",
+					err),
+			}
+			resp, err = btcjson.MarshalResponse("2.0", nil, nil, jsonErr)
```
Actually using `"2.0"` here is fine, as long as it is validated in `btcjson.MarshalResponse`. You can refer to my previous comments.

jakesyl replied:
See https://github.com/btcsuite/btcd/pull/1583#discussion_r533573606, but I agree this should be consistent

jakesyl replied:
Fixed in https://github.com/btcsuite/btcd/pull/1583/commits/50200052a48c6effb6187d7ef26730e760e8416a. We'll continue the discussion on wether this type heavy approach is correct above

onyb reviewed (changes requested):
I think I have carefully reviewed most of the PR but may come back at it again later.

Can you please add some batch JSON-RPC tests to `integration/rpcserver_test.go`?

jcvernaleo commented:
Looking pretty good.  Needs rebase, comments from @onyb addressed, and then squash.

onyb commented on rpcclient/chain.go:
```diff
@@ -1224,3 +1238,24 @@ func (c *Client) GetBlockStatsAsync(hashOrHeight interface{}, stats *[]string) F
+	if err != nil {
+		return nil, err
+	}
+	var arr []IndividualBulkResult
```
`res` may have a different structure. Example:

```
{"result":null,"error":{"code":-32700,"message":"Parse error"},"id":null}
```

jakesyl replied:
wouldn't this be: `[{"result":null,"error":{"code":-32700,"message":"Parse error"},"id":null}]`?

onyb replied:
I don't remember how I encountered it, but it was during some tests. I can try to reproduce it again.

onyb commented on rpcclient/chain.go:
```diff
@@ -1184,6 +1184,20 @@ func (c *Client) GetCFilterHeader(blockHash *chainhash.Hash,
+//from a bulk json rpc api
+type IndividualBulkResult struct {
+	Result interface{} `json:"result"`
+	Error  string      `json:"error"`
```
I think we should use the `*btcjson.RPCError` type here.

jakesyl commented on btcjson/jsonrpc.go:
```diff
@@ -9,6 +9,32 @@ import (
 	"fmt"
 )
 
+type RpcVersion string
+
+const (
+	// version 1 of rpc
+	RpcVersion1 RpcVersion = RpcVersion("1.0")
+	// version 2 of rpc
+	RpcVersion2 RpcVersion = RpcVersion("2.0")
+)
+
+var validRpcVersions = []RpcVersion{RpcVersion1, RpcVersion2}
+
+// check if the rpc version is a valid version
+func (r RpcVersion) IsValid() bool {
+	for _, version := range validRpcVersions {
+		if version == r {
+			return true
+		}
+	}
+	return false
+}
+
+// cast rpc version to a string
+func (r RpcVersion) String() string {
+	return string(r)
+}
```
@onyb This was done to address @torkelrogstad's comment here https://github.com/btcsuite/btcd/pull/1503#discussion_r354181602. I'm happy to change it back, but it seemed like typed constants were preferred.

I've added the golint reccomendation

onyb replied:
It's not so important. Let's keep it this way - at least it improves readability.

jakesyl commented on btcjson/jsonrpc.go:
```diff
@@ -67,21 +93,81 @@ func IsValidIDType(id interface{}) bool {
+	}
+
+	return nil
+}
```
This looks much better, I've mentioned this elsewhere, but if we start doing this it should probably be mentioned in the style guide

jakesyl replied:
btw - had to make one change. rawParams has to be initialized as empty (not null) for this to work correctly

jakesyl commented on btcjson/jsonrpc.go:
```diff
@@ -98,51 +184,63 @@ func NewRequest(id interface{}, method string, params []interface{}) (*Request,
 // response to send over the wire with the MarshalResponse function.
-func NewResponse(id interface{}, marshalledResult []byte, rpcErr *RPCError) (*Response, error) {
+func NewResponse(rpcVersion RpcVersion, id interface{}, marshalledResult []byte, rpcErr *RPCError) (*Response, error) {
+	if rpcVersion != RpcVersion2 && rpcVersion != RpcVersion1 {
+		str := fmt.Sprintf("rpcversion '%s' is invalid", rpcVersion)
+		return nil, makeError(ErrInvalidType, str)
+	}
```
fixed in 9dcd6135

jakesyl commented on btcjson/jsonrpc.go:
```diff
@@ -98,51 +184,63 @@ func NewRequest(id interface{}, method string, params []interface{}) (*Request,
+	if rpcVersion != RpcVersion2 && rpcVersion != RpcVersion1 {
+		str := fmt.Sprintf("rpcversion '%s' is invalid", rpcVersion)
+		return nil, makeError(ErrInvalidType, str)
+	}
```
fixed in 9dcd613

jakesyl commented on rpcclient/examples/bitcoincorehttpbulk/main.go:
```diff
@@ -0,0 +1,50 @@
+
+package main
+
+import (
+	"fmt"
+	"github.com/btcsuite/btcd/rpcclient"
+	"log"
+)
```
huh - I thought goclean would catch this.

jakesyl commented on rpcclient/examples/bitcoincorehttpbulk/main.go:
```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2014-2019 The btcsuite developers
```
tbh - I think it might make sense to do this in a separate PR. The copyrights are kind of a mess:

![image](https://user-images.githubusercontent.com/6392429/100785976-7181fd00-33df-11eb-83e9-8b8a50858abd.png)


![image](https://user-images.githubusercontent.com/6392429/100785986-75ae1a80-33df-11eb-8346-50399bff6a53.png)

![image](https://user-images.githubusercontent.com/6392429/100785995-7a72ce80-33df-11eb-9071-5073f28a709e.png)


I'm not sure if that's because these were seperate reops that got merged together. @jcvernaleo might be more familiar with the history but it looks like it's just messy rather than any nuance: https://github.com/btcsuite/fastsha256/pull/6#issuecomment-239336912

jakesyl commented on rpcclient/examples/bitcoincorehttpbulk/main.go:
```diff
@@ -0,0 +1,50 @@
+	}
+	// Notice the notification parameter is nil since notifications are
+	// not supported in HTTP POST mode.
+	client, err := rpcclient.New(connCfg, nil)
```
Why would it be backwards incompatible? Doesn't batch default to false

jakesyl commented on rpcclient/infrastructure.go:
```diff
@@ -1524,3 +1557,65 @@ func (c *Client) BackendVersion() (BackendVersion, error) {
 	return *c.backendVersion, nil
 }
+
+// Batch makes batch requests
```
I made an attempt at fixing this here: 9c315730

Documentation always be improved so let me know if I'm missing anything too obvious here

jakesyl commented on rpcserver.go:
```diff
@@ -3861,7 +3865,12 @@ handled:
+	}
+
 	parsedCmd.id = request.ID
 	parsedCmd.method = request.Method
```
fixed in 9c315730

jakesyl commented on rpcclient/infrastructure.go:
```diff
@@ -163,6 +163,10 @@ type Client struct {
 	// disconnected indicated whether or not the server is disconnected.
 	disconnected bool
 
+	// wether or not to batch requests, false unless changed by Bulk()
```
fixed in 9c315730

jakesyl commented on rpcclient/infrastructure.go:
```diff
@@ -163,6 +163,10 @@ type Client struct {
 	// disconnected indicated whether or not the server is disconnected.
 	disconnected bool
 
+	// wether or not to batch requests, false unless changed by Bulk()
```
fixed in 9c315730

jakesyl reviewed (commented):
Needs rebase

jakesyl commented:
Alright, so I think I've addressed all the comments here besides rebasing and [this](https://github.com/btcsuite/btcd/pull/1583#issuecomment-683310019) ongoing debate. Huge thanks to @onyb for the comprehensive review and let me know what else here needs cleanup

onyb commented on integration/rpcserver_test.go:
```diff
@@ -10,6 +10,7 @@ package integration
 import (
 	"bytes"
 	"fmt"
+	"github.com/btcsuite/btcd/rpcclient"
```
needs a `goimport`.

onyb commented on integration/rpcserver_test.go:
```diff
@@ -94,10 +95,45 @@ func testGetBlockHash(r *rpctest.Harness, t *testing.T) {
+		t.Fatal(err)
+	}
+
+BLOCKRESULT:
+	for _, block := range futureBlockResults {
+		msgBlock, err := block.Receive()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		blockHash := msgBlock.Header.BlockHash()
+		for _, hash := range generatedBlockHashes {
+			if blockHash.IsEqual(hash) {
+				continue BLOCKRESULT
+			}
+		}
+		t.Fatalf("expected hash %s  to be in generated hash list", blockHash)
+	}
```
Can we rewrite this using a helper function for checking if a received blockhash is part of the generated list? The goto statement makes it a bit hard to understand. Here's an example below (not tested):

```suggestion
    isKnownBlockHash := func(hash string) bool {
        for _, h := range generatedBlockHashes {
            if hash == h {
                return true
            }
        }

        return false
    }

    for _, block := range futureBlockResults {
        msgBlock, err := block.Receive()
        if err != nil {
	    t.Fatal(err)
        }

        blockHash := msgBlock.Header.BlockHash()
        if !isKnownBlockHash(blockHash) {
            t.Fatalf("expected hash %s  to be in generated hash list", blockHash)
        }
    }
```

onyb commented on integration/rpctest/rpc_harness.go:
```diff
@@ -289,8 +290,8 @@ func (h *Harness) SetUp(createTestChain bool, numMatureOutputs uint32) error {
 func (h *Harness) tearDown() error {
-	if h.Node != nil {
-		h.Node.Shutdown()
+	if h.Client != nil {
```
Missing a shutdown for the batch client?

onyb commented on btcjson/jsonrpc.go:
```diff
@@ -9,6 +9,32 @@ import (
 	"fmt"
 )
 
+type RPCVersion string
```
For all exported variables, constants, and types, let's write standard Go comments like:

```
// RPCVersion is a type to indicate RPC versions.
```

onyb commented on rpcwebsocket.go:
```diff
@@ -1323,153 +1323,435 @@ out:
+			batchedRequest = true
 		}
 
-		// The JSON-RPC 1.0 spec defines that notifications must have their "id"
```
It seems a lot of code on the left-side has been removed. Can you please explain this part?

jakesyl replied:
A lot of the validations been moved further down the tree since it only applied to 1.0 rpc

onyb reviewed (changes requested):
Adding my last set of review comments. They are mostly minor ones. Looking really good so far, and I can see the light at the end of the tunnel.

While you address the remaining comments, I'll try to take some time and do some QA.

jakesyl commented:
should be all set with those changes. Let me know if you need a little more clarity here - https://github.com/btcsuite/btcd/pull/1583#discussion_r542959139. I was a little unclear on the block you were referring too

onyb commented on rpcclient/infrastructure.go:
```diff
@@ -1534,3 +1620,60 @@ func (c *Client) BackendVersion() (BackendVersion, error) {
+// creates a response channel to receive the response
+func (c *Client) Send() error {
+	result, err := c.sendAsync().Receive()
+
```
I think we need to do something like this here, in order to clear the `batchList` even in case of an error:

```
defer func() {
    c.batchList = list.New()
}
```

What do you think?

onyb replied:
If `Send()` is called on an empty `batchList`, it should be a noop.

```
if c.batchList.Len() == 0 {
    return nil
}
```

The above needs to be called **before** `result, err := c.sendAsync().Receive()` is done to prevent a JSON marshalling error.

onyb commented on rpcserver.go:
```diff
@@ -4059,80 +4125,165 @@ func (s *rpcServer) jsonRPCRead(w http.ResponseWriter, r *http.Request, isAdmin
-		//
-		// RPC quirks can be enabled by the user to avoid compatibility issues
-		// with software relying on Core's behavior.
-		if request.ID == nil && !(cfg.RPCQuirks && request.Jsonrpc == "") {
-			return
```
Is it intentional that these lines have been removed?

jakesyl replied:
You're right. I've added this back to the json rpc 1.0 handler

jakesyl commented on integration/rpcserver_test.go:
```diff
@@ -10,27 +10,30 @@ package integration
 import (
 	"bytes"
 	"fmt"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
```
do we have a test case for this? Is this not showing up in the golint because of the no build flag?

jakesyl commented:
btw - did a little load testing on this. Works great 
![image](https://user-images.githubusercontent.com/6392429/106875676-ef46db00-66a4-11eb-95f0-94954e1f8561.png)

jakesyl commented:
Hey, what do we need to get this in?

onyb reviewed (approved):
I'm unable to resolve conversations on this PR, but it seems all review comments have been addressed/answered. I did some quick tests on my side, and it works great.

Congrats on making it to the end! 🚀

jcvernaleo reviewed (approved):
OK

This is awesome.  Nice work.

And good job from all the reviewers too!
